<h3>Importing required libraries</h3>

In [78]:
import numpy as np # library to handle data in a vectorized manner
import geopy
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
import requests
print('Libraries imported.')

Libraries imported.


<h3>The data with postal codes, borough, neighborhood and location coordinates is read from the file</h3>

In [79]:
df = pd.read_csv('scrapped_data_coordinates.csv')
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

In [80]:
df.shape

(103, 5)

<h3>Obtaining the location coordinates of Toronto</h3>

In [49]:
address = 'Toronto'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


<h3>Plotting all the data on the map</h3>

In [81]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)
cnt = 1
# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    
map_toronto

<h3>Filtering the original data such only boroughs containing Toronto are obtained</h3>

In [82]:
toronto_df = df[df['Borough'].str.contains("Toronto")]
toronto_df

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                   North Toronto West, Lawrence Park  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

<h3>Connecting to the foursquare app using the personal credentials</h3>

In [83]:
CLIENT_ID = 'RHZBRTOAICXAP0RRIOX12HLOBSNMAOLMHJUFKLMC1BU1NQ1Z' # your Foursquare ID
CLIENT_SECRET = 'AM0HLEVROB5JI2QMXXOIJPQK5C3WXNY1OMIRRYRNZM3MZGXY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: RHZBRTOAICXAP0RRIOX12HLOBSNMAOLMHJUFKLMC1BU1NQ1Z
CLIENT_SECRET:AM0HLEVROB5JI2QMXXOIJPQK5C3WXNY1OMIRRYRNZM3MZGXY


<h3>Obtaining the 100(max) venues nearby for each postal code location in the radius of 500 units using foursquare API endpoint</h3>

In [57]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_df['Latitude'],toronto_df['Longitude'],toronto_df['Postal Code'],toronto_df['Borough'],toronto_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))
    

<h3>Result dataframe containing the venues nearby for each record</h3>

In [84]:
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Postal Code', 'Borough', 'Neighborhood', 'Borough Latitude', 'Borough Longitude', 'Venue Name', 'Venue Latitude', 'Venue Longitude', 'Venue Category']

print(venues_df.shape)
venues_df.head()

(1620, 9)


Postal Code           Borough               Neighborhood  Borough Latitude  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront          43.65426   
1         M5A  Downtown Toronto  Regent Park, Harbourfront          43.65426   
2         M5A  Downtown Toronto  Regent Park, Harbourfront          43.65426   
3         M5A  Downtown Toronto  Regent Park, Harbourfront          43.65426   
4         M5A  Downtown Toronto  Regent Park, Harbourfront          43.65426   

   Borough Longitude                Venue Name  Venue Latitude  \
0         -79.360636          Roselle Desserts       43.653447   
1         -79.360636             Tandem Coffee       43.653559   
2         -79.360636    Cooper Koo Family YMCA       43.653249   
3         -79.360636       Body Blitz Spa East       43.654735   
4         -79.360636  Dominion Pub and Kitchen       43.656919   

   Venue Longitude       Venue Category  
0       -79.362017               Bakery  
1       -79.361809          Coffee Shop  
2       -79.358008  Distribution Center  
3       -79.359874                  Spa  
4       -79.358967                  Pub

<h3>Below shows the number of venues for each toronto location</h3>

In [85]:
venues_df.groupby(["Postal Code", "Borough", "Neighborhood"]).count()

Borough Latitude  \
Postal Code Borough          Neighborhood                                                           
M4E         East Toronto     The Beaches                                                        5   
M4K         East Toronto     The Danforth West, Riverdale                                      43   
M4L         East Toronto     India Bazaar, The Beaches West                                    24   
M4M         East Toronto     Studio District                                                   40   
M4N         Central Toronto  Lawrence Park                                                      4   
M4P         Central Toronto  Davisville North                                                   9   
M4R         Central Toronto  North Toronto West, Lawrence Park                                 20   
M4S         Central Toronto  Davisville                                                        33   
M4T         Central Toronto  Moore Park, Summerhill East                                        3   
M4V         Central Toronto  Summerhill West, Rathnelly, South Hill, Forest ...                16   
M4W         Downtown Toronto Rosedale                                                           4   
M4X         Downtown Toronto St. James Town, Cabbagetown                                       46   
M4Y         Downtown Toronto Church and Wellesley                                              75   
M5A         Downtown Toronto Regent Park, Harbourfront                                         44   
M5B         Downtown Toronto Garden District, Ryerson                                         100   
M5C         Downtown Toronto St. James Town                                                    78   
M5E         Downtown Toronto Berczy Park                                                       56   
M5G         Downtown Toronto Central Bay Street                                                64   
M5H         Downtown Toronto Richmond, Adelaide, King                                          92   
M5J         Downtown Toronto Harbourfront East, Union Station, Toronto Islands                100   
M5K         Downtown Toronto Toronto Dominion Centre, Design Exchange                         100   
M5L         Downtown Toronto Commerce Court, Victoria Hotel                                   100   
M5N         Central Toronto  Roselawn                                                           1   
M5P         Central Toronto  Forest Hill North & West, Forest Hill Road Park                    6   
M5R         Central Toronto  The Annex, North Midtown, Yorkville                               21   
M5S         Downtown Toronto University of Toronto, Harbord                                    34   
M5T         Downtown Toronto Kensington Market, Chinatown, Grange Park                         62   
M5V         Downtown Toronto CN Tower, King and Spadina, Railway Lands, Harb...                16   
M5W         Downtown Toronto Stn A PO Boxes                                                    96   
M5X         Downtown Toronto First Canadian Place, Underground city                           100   
M6G         Downtown Toronto Christie                                                          17   
M6H         West Toronto     Dufferin, Dovercourt Village                                      16   
M6J         West Toronto     Little Portugal, Trinity                                          43   
M6K         West Toronto     Brockton, Parkdale Village, Exhibition Place                      24   
M6P         West Toronto     High Park, The Junction South                                     25   
M6R         West Toronto     Parkdale, Roncesvalles                                            15   
M6S         West Toronto     Runnymede, Swansea                                                38   
M7A         Downtown Toronto Queen's Park, Ontario Provincial Government                       34   
M7Y         East Toronto     Business rep

In [86]:
print('There are {} uniques categories.'.format(len(venues_df['Venue Category'].unique())))

There are 237 uniques categories.


<h3>So we know that there are total of 237 unique venue categories. Obtaining one hot encoding for all the venue categories at each postal code location</h3>

In [87]:
toronto_onehot = pd.get_dummies(venues_df[['Venue Category']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_onehot['Postal Code'] = venues_df['Postal Code'] 
toronto_onehot['Borough'] = venues_df['Borough'] 
toronto_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_onehot.columns[-3:]) + list(toronto_onehot.columns[:-3])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(1620, 240)


Postal Code           Borough              Neighborhoods  Afghan Restaurant  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
1         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
2         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
3         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
4         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       1     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Rental / Bike Share  Bistro  \
0           0                   0                         0       0   
1           0                   0                         0       0   
2           0                   0                         0       0   
3           0                   0                         0       0   
4           0                   0                         0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3             

In [88]:
toronto_grouped = toronto_onehot.groupby(["Postal Code", "Borough", "Neighborhoods"]).mean().reset_index()

print(toronto_grouped.shape)
toronto_grouped

(39, 240)


Postal Code           Borough  \
0          M4E      East Toronto   
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
4          M4N   Central Toronto   
5          M4P   Central Toronto   
6          M4R   Central Toronto   
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
10         M4W  Downtown Toronto   
11         M4X  Downtown Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
16         M5E  Downtown Toronto   
17         M5G  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
22         M5N   Central Toronto   
23         M5P   Central Toronto   
24         M5R   Central Toronto   
25         M5S  Downtown Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
28         M5W  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
34         M6P      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighborhoods  Afghan Restaurant  \
0                                         The Beaches           0.000000   
1                        The Danforth West, Riverdale           0.000000   
2                      India Bazaar, The Beaches West           0.000000   
3                                     Studio District           0.000000   
4                                       Lawrence Park           0.000000   
5                                    Davisville North           0.000000   
6                   North Toronto West, Lawrence Park           0.000000   
7                                          Davisville           0.000000   
8                         Moore Park, Summerhill East           0.000000   
9   Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
10                                           Rosedale           0.000000   
11                        St. James Town, Cabbagetown           0.000000   
12                               Church and Wellesley           0.013333   
13                          Regent Park, Harbourfront           0.000000   
14                           Garden District, Ryerson           0.000000   
15                                     St. James Town           0.000000   
16                                        Berczy Park           0.000000   
17                                 Central Bay Street           0.000000   
18                           Richmond, Adelaide, King           0.000000   
19  Harbourfront East, Union Station, Toronto Islands           0.000000   
20           Toronto Dominion Centre, Design Exchange           0.000000   
21                     Commerce Court, Victoria Hotel           0.000000   
22                                           Roselawn           0.000000   
23    Forest Hill North & West, Forest Hill Road Park           0.000000   
24                The Annex, North Midtown, Yorkville           0.000000   
25                     University of Toronto, Harbord           0.000000   
26          Kensington Market, Chinatown, Grange Park           0.000000   
27  CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
28                                     Stn A PO Boxes           0.000000   
29             First Canadian Place, Underground city           0.000000   
30                                           Christie           0.000000   
31                       Dufferin, Dovercourt Village           0.000000   
32                           Little Portugal, Trinity 

<h3>Reducing the data dimensionality by selecting only top 10 venues for each postal code</h3>

In [90]:
#Now let's create the new dataframe and display the top 10 venues for each PostalCode.
num_top_venues = 10
indicators = ['st', 'nd', 'rd']
# create columns according to number of top venues
areaColumns = ['Postal Code', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']
neighbourhoods_venues_sorted['Borough'] = toronto_grouped['Borough']
neighbourhoods_venues_sorted['Neighborhoods'] = toronto_grouped['Neighborhoods']

for ind in np.arange(toronto_grouped.shape[0]):
    row_categories = toronto_grouped.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighbourhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

# neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
print(neighbourhoods_venues_sorted.shape)
neighbourhoods_venues_sorted

(39, 13)


Postal Code           Borough  \
0          M4E      East Toronto   
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
4          M4N   Central Toronto   
5          M4P   Central Toronto   
6          M4R   Central Toronto   
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
10         M4W  Downtown Toronto   
11         M4X  Downtown Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
16         M5E  Downtown Toronto   
17         M5G  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
22         M5N   Central Toronto   
23         M5P   Central Toronto   
24         M5R   Central Toronto   
25         M5S  Downtown Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
28         M5W  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
34         M6P      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighborhoods 1st Most Common Venue  \
0                                         The Beaches     Health Food Store   
1                        The Danforth West, Riverdale      Greek Restaurant   
2                      India Bazaar, The Beaches West                  Park   
3                                     Studio District                  Café   
4                                       Lawrence Park                  Park   
5                                    Davisville North                  Park   
6                   North Toronto West, Lawrence Park        Clothing Store   
7                                          Davisville           Pizza Place   
8                         Moore Park, Summerhill East                  Park   
9   Summerhill West, Rathnelly, South Hill, Forest...                   Pub   
10                                           Rosedale                  Park   
11                        St. James Town, Cabbagetown           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                          Regent Park, Harbourfront           Coffee Shop   
14                           Garden District, Ryerson        Clothing Store   
15                                     St. James Town                  Café   
16                                        Berczy Park           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
18                           Richmond, Adelaide, King           Coffee Shop   
19  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
20           Toronto Dominion Centre, Design Exchange           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
22                                           Roselawn                Garden   
23    Forest Hill North & West, Forest Hill Road Park                 Trail   
24                The Annex, North Midtown, Yorkville                  Café   
25                     University of Toronto, Harbord                  Café   
26          Kensington Market, Chinatown, Grange Park                  Café   
27  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
28                                     Stn A PO Boxes           Coffee Shop   
29             First Canadian Place, Underground city           Coffee Shop   
30                                           Christie         Grocery Store   
31                       Dufferin,

In [91]:
#Run k-means to cluster the Toronto areas into 5 clusters.
# set number of clusters
kclusters=5
toronto_grouped_clustering=toronto_grouped.drop(["Postal Code", "Borough", "Neighborhoods"], 1)
# run k-means clustering
kmeans=KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 4, 1, 1, 1, 3, 1])

In [92]:
toronto_merged=toronto_df.copy()

# add clustering labels
toronto_merged["Cluster Labels"] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.drop(["Borough", "Neighborhoods"], 1).set_index("Postal Code"), on="Postal Code")

print(toronto_merged.shape)
toronto_merged.head() # check the last columns!

(39, 16)


Postal Code           Borough                                 Neighborhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               1           Coffee Shop   
4   43.662301 -79.389494               1           Coffee Shop   
9   43.657162 -79.378937               1        Clothing Store   
15  43.651494 -79.375418               1                  Café   
19  43.676357 -79.293031               4     Health Food Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park                   Pub                Bakery   
4       Sushi Restaurant                 Diner           Yoga Studio   
9            Coffee Shop                  Café        Cosmetics Shop   
15           Coffee Shop             Gastropub            Restaurant   
19          Neighborhood      Asian Restaurant                   Pub   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                        Café               Theater            Restaurant   
4              Discount Store                   Bar              Beer Bar   
9   Middle Eastern Restaurant   Japanese Restaurant       Bubble Tea Shop   
15        American Restaurant          Cocktail Bar        Lingerie Store   
19                      Trail      Department Store          Dessert Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2         Breakfast Spot           Yoga Studio             Shoe Store  
4          Smoothie Shop      Sculpture Garden         Sandwich Place  
9     Italian Restaurant               Theater       Ramen Restaurant  
15   Moroccan Restaurant              Creperie                  Hotel  
19    Dim Sum Restaurant                 Diner            Yoga Studio

In [93]:
# sort the results by Cluster Labels
print(toronto_merged.shape)
toronto_merged.sort_values(["Cluster Labels"], inplace=True)
toronto_merged

(39, 16)


Postal Code           Borough  \
73          M4R   Central Toronto   
2           M5A  Downtown Toronto   
68          M5P   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
67          M4P   Central Toronto   
99          M4Y  Downtown Toronto   
62          M5N   Central Toronto   
54          M4M      East Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
61          M4N   Central Toronto   
31          M6H      West Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
100         M7Y      East Toronto   
69          M6P      West Toronto   
30          M5H  Downtown Toronto   
36          M5J  Downtown Toronto   
19          M4E      East Toronto   

                                          Neighborhood   Latitude  Longitude  \
73                   North Toronto West, Lawrence Park  43.715383 -79.405678   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678   
75                              Parkdale, Roncesvalles  43.648960 -79.456325   
79                                          Davisville  43.704324 -79.388790   
80                      University of Toronto, Harbord  43.662696 -79.400049   
81                                  Runnymede, Swansea  43.651571 -79.484450   
83                         Moore Park, Summerhill East  43.689574 -79.383160   
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
86   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
87   CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
91                                            Rosedale  43.679563 -79.377529   
92                                      Stn A PO Boxes  43.646435 -79.374846   
96                         St. James Town, Cabbagetown  43.667967 -79.367675   
97              First Canadian Place, Underground city  43.648429 -79.382280   
67                                    Davisville North  43.712751 -79.390197   
99                                Church and Wellesley  43.665860 -79.383160   
62                                            Roselawn  43.711695 -79.416936   
54                                     Studio District  43.659526 -79.340923   
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
9                             Garden District, Ryerson  43.657162 -79.378937   
15                                      St. James Town  43.651494 -79.375418   
20                                         Berczy Park  43.644771 -79.373306   
24                                  Central Bay Street  43.657952 -79.387383   
25                                            Christie  43.669542 -79.422564   
61                                       Lawrence Park  43.728020 -79.388790   
31                        Dufferin, Dovercourt Village  43.669005 -79.442259   
37                            Little Portugal, Trinity  43.647927 -79.419750   
41                        The Danforth West, Riverdale  43.679557 -79.352188   
42            Toronto Dominion Centre, Desi

In [94]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postal Code'], toronto_merged['Borough'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [95]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
73  Central Toronto               0        Clothing Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
73           Coffee Shop           Yoga Studio  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
73   Rental Car Location    Mexican Restaurant            Shoe Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
73            Restaurant                  Café    Sporting Goods Shop

In [96]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto               1           Coffee Shop   
68    Central Toronto               1                 Trail   
74    Central Toronto               1                  Café   
75       West Toronto               1             Gift Shop   
79    Central Toronto               1           Pizza Place   
80   Downtown Toronto               1                  Café   
81       West Toronto               1           Coffee Shop   
83    Central Toronto               1                  Park   
84   Downtown Toronto               1                  Café   
86    Central Toronto               1                   Pub   
87   Downtown Toronto               1       Airport Service   
91   Downtown Toronto               1                  Park   
92   Downtown Toronto               1           Coffee Shop   
96   Downtown Toronto               1           Coffee Shop   
97   Downtown Toronto               1           Coffee Shop   
67    Central Toronto               1                  Park   
99   Downtown Toronto               1           Coffee Shop   
62    Central Toronto               1                Garden   
54       East Toronto               1                  Café   
4    Downtown Toronto               1           Coffee Shop   
9    Downtown Toronto               1        Clothing Store   
15   Downtown Toronto               1                  Café   
20   Downtown Toronto               1           Coffee Shop   
24   Downtown Toronto               1           Coffee Shop   
25   Downtown Toronto               1         Grocery Store   
61    Central Toronto               1                  Park   
31       West Toronto               1                Bakery   
37       West Toronto               1                   Bar   
41       East Toronto               1      Greek Restaurant   
42   Downtown Toronto               1           Coffee Shop   
43       West Toronto               1                  Café   
47       East Toronto               1                  Park   
48   Downtown Toronto               1           Coffee Shop   
100      East Toronto               1                Garden   

             2nd Most Common Venue 3rd Most Common Venue  \
2                             Park                   Pub   
68                            Park              Bus Line   
74                  Sandwich Place           Coffee Shop   
75                  Breakfast Spot         Movie Theater   
79                    Dessert Shop        Sandwich Place   
80             Japanese Restaurant                   Bar   
81                            Café           Pizza Place   
83                      Restaurant            Playground   
84   Vegetarian / Vegan Restaurant                Bakery   
86                     Coffee Shop           Supermarket   
87                  Airport Lounge      Airport Terminal   
91                      Playground                 Trail   
92                            Café    Italian Restaurant   
96                            Café                Bakery   
97                            Café            Restaurant   
67                     Pizza Place        Breakfast Spot   
99                Sushi Restaurant   Japanese Restaurant   
62                     Yoga Studio          Dance Studio   
54                     Coffee Shop             Gastropub   
4                 Sushi Restaurant                 Diner   
9                      Coffee Shop                  Café   
15                     Coffee Shop             Gastropub   
20                    Cocktail Bar              Beer Bar   
24                  Sandwich Place    Italian Restaurant   
25                            Café                  Park   
61                        Bus Line           Swim School   
31                        Pharmacy             Pet Store   
37                            Café           Coffee Shop   
41                     Coffee Shop    Italian Restaurant   
42      

In [97]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
69  West Toronto               2                  Café    Mexican Restaurant   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
69  Furniture / Home Store       Thai Restaurant            Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
69             Speakeasy           Flea Market                   Bar   

   9th Most Common Venue 10th Most Common Venue  
69                Bakery   Fast Food Restaurant

In [98]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
30  Downtown Toronto               3           Coffee Shop   
36  Downtown Toronto               3           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
30                  Café            Restaurant                   Gym   
36              Aquarium                  Café                 Hotel   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
30       Thai Restaurant                 Hotel         Deli / Bodega   
36        Scenic Lookout               Brewery   Sporting Goods Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
30      Sushi Restaurant             Bookstore            Pizza Place  
36            Restaurant   Fried Chicken Joint                    Bar

In [99]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19  East Toronto               4     Health Food Store          Neighborhood   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19      Asian Restaurant                   Pub                 Trail   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19      Department Store          Dessert Shop    Dim Sum Restaurant   

   9th Most Common Venue 10th Most Common Venue  
19                 Diner            Yoga Studio

Observations:
Most of the neighborhoods fall into Cluster 1 which are mostly business areas with cafe, restaurants, supermarkets etc. Cluster 0 and cluster 4 comprise of clothing store and health food store as primary venues respectively. Cluster 2 has only Cafe as the most common venue where as cluster 3 has coffee shop as the most common venue